From 80174e245b0e525c660bcbcc5a8279548e017b4b Mon Sep 17 00:00:00 2001
From: Adrien Destugues <pulkomandy@pulkomandy.tk>
Date: Tue, 15 Aug 2017 13:09:47 +0200
Subject: Port Korli's inital 1.1.0 patch

- Build fixes (some were upstreamed)
- Disable mem_sec because we don't have mlock

diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
index 39b89e7..1179d3f 100644
--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -1764,6 +1764,32 @@ sub vms_info {
         shared_extension => ".so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
         ranlib           => "$ENV{'RANLIB'}",
     },
+    "haiku-common" => {
+        template         => 1,
+        cc               => "cc",
+        cflags           => "-DL_ENDIAN -Wall",
+        debug_cflags     => "-g -O0",
+        release_cflags   => "-O2",
+        thread_cflag     => "-D_REENTRANT",
+        sys_id           => "HAIKU",
+        lflags           => "-lnetwork",
+        perlasm_scheme   => "elf",
+        dso_scheme       => "dlfcn",
+        shared_target    => "haiku-shared",
+        shared_cflag     => "-fPIC",
+        shared_ldflag    => "-shared",
+        shared_extension => ".so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
+    },
+    "haiku-x86" => {
+        inherit_from     => [ "haiku-common", asm("x86_elf_asm") ],
+        release_cflags   => add(" ", "-fomit-frame-pointer"),
+        bn_ops           => "BN_LLONG",
+    },
+    "haiku-x86_64" => {
+        inherit_from     => [ "haiku-common", asm("x86_64_asm") ],
+        cflags           => add(" ", "-m64"),
+        bn_ops           => "SIXTY_FOUR_BIT_LONG",
+    },
 
     ##### VMS
     "vms-generic" => {
diff --git a/Makefile.shared b/Makefile.shared
index 098e1ec..f964897 100644
--- a/Makefile.shared
+++ b/Makefile.shared
@@ -564,11 +564,11 @@ symlink.hpux:
 symlink.cygwin symlink.alpha-osf1 symlink.tru64 symlink.tru64-rpath:
 
 # Compatibility targets
-link_dso.bsd-gcc-shared link_dso.linux-shared link_dso.gnu-shared: link_dso.gnu
+link_dso.bsd-gcc-shared link_dso.linux-shared link_dso.gnu-shared link_dso.haiku-shared: link_dso.gnu
 link_shlib.bsd-gcc-shared: link_shlib.linux-shared
-link_shlib.gnu-shared: link_shlib.gnu
-link_app.bsd-gcc-shared link_app.linux-shared link_app.gnu-shared: link_app.gnu
-symlink.bsd-gcc-shared symlink.bsd-shared symlink.linux-shared symlink.gnu-shared: symlink.gnu
+link_shlib.gnu-shared link_shlib.haiku-shared: link_shlib.gnu
+link_app.bsd-gcc-shared link_app.linux-shared link_app.gnu-shared linka_app.haiku-shared: link_app.gnu
+symlink.bsd-gcc-shared symlink.bsd-shared symlink.linux-shared symlink.gnu-shared symlink.haiku-shared: symlink.gnu
 link_dso.bsd-shared: link_dso.bsd
 link_shlib.bsd-shared: link_shlib.bsd
 link_app.bsd-shared: link_app.bsd
diff --git a/crypto/include/internal/cryptlib.h b/crypto/include/internal/cryptlib.h
index f3ec9b6..b3c80b4 100644
--- a/crypto/include/internal/cryptlib.h
+++ b/crypto/include/internal/cryptlib.h
@@ -42,7 +42,7 @@ DEFINE_LHASH_OF(MEM);
 # ifndef OPENSSL_SYS_VMS
 #  define X509_CERT_AREA          OPENSSLDIR
 #  define X509_CERT_DIR           OPENSSLDIR "/certs"
-#  define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
+#  define X509_CERT_FILE          OPENSSLDIR "/CARootCertificates.pem"
 #  define X509_PRIVATE_DIR        OPENSSLDIR "/private"
 #  define CTLOG_FILE              OPENSSLDIR "/ct_log_list.cnf"
 # else
diff --git a/crypto/mem_sec.c b/crypto/mem_sec.c
index 664b4ad..3020e3a 100644
--- a/crypto/mem_sec.c
+++ b/crypto/mem_sec.c
@@ -25,7 +25,9 @@
 #include <string.h>
 
 #if defined(OPENSSL_SYS_LINUX) || defined(OPENSSL_SYS_UNIX)
+#ifndef OPENSSL_SYS_HAIKU
 # define IMPLEMENTED
+#endif
 # include <stdlib.h>
 # include <assert.h>
 # include <unistd.h>
diff --git a/e_os.h b/e_os.h
index 559bf90..36b7963 100644
--- a/e_os.h
+++ b/e_os.h
@@ -521,6 +521,13 @@ struct servent *getservbyname(const char *name, const char *proto);
 # endif
 /* end vxworks */
 
+/* haiku */
+# if defined(OPENSSL_SYS_HAIKU)
+#  include <sys/select.h>
+#  include <sys/time.h>
+#endif
+/* end haiku */
+
 #define OSSL_NELEM(x)    (sizeof(x)/sizeof(x[0]))
 
 #ifdef  __cplusplus
-- 
2.7.0

